Sources/Datadog/Tracing/Utils/URLSessionSwizzler.swift

```diff
@@ -0,0 +1,104 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+
+// TODO: RUMM-300 plug DDTracer into swizzled methods
+
+internal class URLSessionSwizzler {
```

Comment:
I really like the way `URLSessionSwizzler` implementation looks like 👌🏅! It's easy to read and clearly shows how to add new swizzlings.

However, this kind of interface:
```swift
static func swizzle() throws
static func unswizzle() throws
static func swizzleDataTaskWithURL() throws
static func swizzleDataTaskWithRequest() throws
static func swizzleDataTaskWithURLCompletionHandler() throws
static func swizzleDataTaskWithRequestCompletionHandler() throws
```
will force us to break the single responsibility of `URLSessionSwizzler` and bloat it with our custom logic - today for `URLRequest` tracing, but later for RUM product features (resources tracing). It will become impossible to test different pieces of logics individually as the `URLSessionSwizzlerTests` will be touching different products (mocking hell 🔥 ).

The problem with testing will only reveal a wider, architectural issue: as adding new features won't be based on composition, it will require touching the existing stuff, thus may break them i.e. the tracing auto-instrumentation may be broken when adding resource tracing for RUM.

Last, but not least - this is the first `*Swizzler` that we add. For other, most likely, we'll follow the same patterns, and the composition problem will be escalated. Single `UIViewControllerSwizzler` will become giant because of implementing all the RUM auto-instrumentation.

Let's make the interface like this:
```swift
class URLSessionSwizzler {
    static func interceptDataTaskWithURL(_ callback: (...) -> Void) throws
    static func interceptDataTaskWithRequest(_ callback: (...) -> Void) throws
    static func interceptDataTaskWithURLCompletionHandler(_ callback: (...) -> Void) throws
    static func interceptDataTaskWithRequestCompletionHandler(_ callback: (...) -> Void) throws
}
```
so it can be composed:
```swift
class URLSessionTracingInterceptor {
    static func initialize() -> URLSessionTracingInterceptor {
        let interceptor = URLSessionTracingInterceptor()
        URLSessionSwizzler.interceptDataTaskWithRequestCompletionHandler(interceptor.handleURLSessionCalll)
        // ...
        return interceptor
    }

    func handleURLSessionCalll(task: URLSessionTask, completion: (Data?, URLResponse?, Error?)) {}
    // ...
}
```
This way:
* we can unit test Swizzling separately in `URLSessionSwizzlerTests`;
* we can unit test tracing instrumentation separately in `URLSessionTracingInterceptorTests`;
* the `initialize()` method can be tested in as a loosely coupled integration test.

WDYT @buranmert ?

Reply:
This is out of scope of this PR, there is another ticket for this.
This PR basically implements the swizzler which doesn't blow up.

We need to talk about how `Tracer` and `RUM` will be used.
I was thinking of passing `SharedTracer` and `SharedRUM` to `swizzle` methods, this will be easier to read than passing blocks.
Passing blocks was necessary in isa swizzling (because of ObjC<->Swift) but now it's not needed.

Reply:
> This is out of scope of this PR, there is another ticket for this.

OK, we can refactor this in separate PR, as I don't see clearly the idea for using `Tracer` for interception.

---

> I was thinking of passing SharedTracer and SharedRUM to swizzle methods, this will be easier to read than passing blocks.

If we use features directly in the swizzled methods, the `*Swizzler` won't follow the SRP, which will result with the problems I explained here:

> (This) will force us to break the single responsibility of URLSessionSwizzler and bloat it with our custom logic - today for URLRequest tracing, but later for RUM product features (resources tracing). It will become impossible to test different pieces of logics individually as the URLSessionSwizzlerTests will be touching different products (mocking hell 🔥 ).

> The problem with testing will only reveal a wider, architectural issue: as adding new features won't be based on composition, it will require touching the existing stuff, thus may break them i.e. the tracing auto-instrumentation may be broken when adding resource tracing for RUM.

Reply:
✅ Injecting blocks is an option
ℹ️ i'm resolving this thread as we will have this discussion in the next PR